### ISSUE TITLE
CMake Error at src/lfortran/CMakeLists.txt:40 (add_library): Cannot find source file: parser/preprocessor.cpp

### ISSUE THREAD

**Issue by barracuda156** [lfortran/lfortran#1212]
```
CMake Error at src/lfortran/CMakeLists.txt:40 (add_library):
  Cannot find source file:

    parser/preprocessor.cpp

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm .h
  .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90 .f95 .f03 .hip .ispc


CMake Error at src/lfortran/CMakeLists.txt:40 (add_library):
  No SOURCES given to target: lfortran_lib
```

It is indeed missing: https://github.com/lfortran/lfortran/tree/main/src/lfortran/parser

**Comment by certik** [comment]
You have to run `./build0.sh`.

**Comment by certik** [comment]
See https://docs.lfortran.org/en/installation/#build-from-git.

**Comment by barracuda156** [comment]
@certik I was actually trying to build from release tarball (and it is broken: when you try running the script, it fails with a complain that it is not a git version).

After fetching the master, it is building, though I got a bunch of weird `bison` warnings `time limit exceeded: 6.000000`. Thankfully, LLVM does not seem to be required.

**Comment by barracuda156** [comment]
Failed in the mid of the build with:
```
In file included from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_lfortran/lfortran/work/lfortran-0.18.0/src/lfortran/parser/tokenizer.h:5,
                 from tokenizer.re:4:
/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_lfortran/lfortran/work/lfortran-0.18.0/src/lfortran/parser/parser_stype.h:97:31: error: static assertion failed
   97 | static_assert(sizeof(YYSTYPE) == sizeof(Vec<AST::ast_t*>));
      |               ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_lfortran/lfortran/work/lfortran-0.18.0/src/lfortran/parser/parser_stype.h:97:31: note: the comparison reduces to '(16 == 12)'
In file included from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_lfortran/lfortran/work/lfortran-0.18.0/src/lfortran/parser/tokenizer.h:5,
                 from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_lfortran/lfortran/work/lfortran-0.18.0/src/lfortran/parser/parser.h:10,
                 from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_lfortran/lfortran/work/lfortran-0.18.0/src/lfortran/parser/preprocessor.h:6,
                 from preprocessor.re:4:
/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_lfortran/lfortran/work/lfortran-0.18.0/src/lfortran/parser/parser_stype.h:97:31: error: static assertion failed
   97 | static_assert(sizeof(YYSTYPE) == sizeof(Vec<AST::ast_t*>));
      |               ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_lang_lfortran/lfortran/work/lfortran-0.18.0/src/lfortran/parser/parser_stype.h:97:31: note: the comparison reduces to '(16 == 12)'
```

**Comment by certik** [comment]
Ah I see, it wasn't clear to me from the report. In that case that is indeed a bug if it doesn't work. We need to look into it. When we moved from GitLab to GitHub, we had to rework the CI and it must have gotten broken, I am sorry about that.

For git version, you have to use Bison 3.4 due to https://github.com/lfortran/lfortran/issues/290, see here for the exact versions of all dependencies that work: https://github.com/lfortran/lfortran/blob/main/environment_linux.yml.

**Comment by barracuda156** [comment]
@certik Thank you! It builds in fact, with default Macports versions of tools. The only issue is that static assert – it is likely wrong for PPC (possibly due to a different size of bool). Once removed, build is successful.

I am rebuilding now with enabled support for json and libfmt. That will likely work.

There is no support for tests with CMake so far? CMakeLists have `enable_testing`, but tests dir is not added, and does not have its CMakeLists.

**Comment by certik** [comment]
Do you have an older Mac that is PowerPC?

The test directory is executed using `./run_tests.py`. We also have `integration_tests` that use cmake, but you can run it with `./run_tests.py` in that directory. Finally, we also have regular `ctest` tests, just run `ctest`, they live in `src/lfortran/tests`.

**Comment by barracuda156** [comment]
> Do you have an older Mac that is PowerPC?

I have several, but yes, half of them PPC.

> The test directory is executed using `./run_tests.py`. We also have `integration_tests` that use cmake, but you can run it with `./run_tests.py` in that directory. Finally, we also have regular `ctest` tests, just run `ctest`, they live in `src/lfortran/tests`.

Ctest didn’t work for me. No tests were enabled. (Well, I am sure it is doable manually, but it would be far easier to have them included into CMakeLists and enabled with -DBUILD_TESTING, so that after the build Ctest can be run directly.)

**Comment by certik** [comment]
> I have several, but yes, half of them PPC.

Excellent. Will you please help us get everything compiling and tests passing out of the box on PPC? 

I don't have access to any PPC machine, but I'll help you answer any questions that you might have.

> Ctest didn’t work for me. No tests were enabled. (Well, I am sure it is doable manually, but it would be far easier to have them included into CMakeLists and enabled with -DBUILD_TESTING, so that after the build Ctest can be run directly.)

They are already included and they are enabled. So what you experienced is a bug. Can you please open up a new issue?

**Comment by barracuda156** [comment]
> Excellent. Will you please help us get everything compiling and tests passing out of the box on PPC?

Yes, sure, I will be happy to.

> I don't have access to any PPC machine, but I'll help you answer any questions that you might have.

Awesome, thanks.

> They are already included and they are enabled. So what you experienced is a bug. Can you please open up a new issue for that?

Let me verify again today to make sure what works and what does not. If it is confirmed not to behave as expected, I will open an issue.

**Comment by certik** [comment]
Awesome, thank you @barracuda156 !

**Comment by barracuda156** [comment]
@certik Okay, tests are working, though very few are activated via Ctest by default:
```
--->  Testing lfortran
Executing:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_lfortran/lfortran/work/lfortran-0.18.0" && ctest test 
Test project /opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_lang_lfortran/lfortran/work/lfortran-0.18.0
    Start 1: test_cwrapper
1/3 Test #1: test_cwrapper ....................   Passed    0.02 sec
    Start 2: test_stacktrace
2/3 Test #2: test_stacktrace ..................   Passed    0.02 sec
    Start 3: test_lfortran
3/3 Test #3: test_lfortran ....................   Passed    0.09 sec

100% tests passed, 0 tests failed out of 3

Total Test time (real) =   0.20 sec
```

**Comment by certik** [comment]
Yes, those are all the tests. The `test_lfortran` has a lot of tests in it.
